src/Shelly.hs

```diff
@@ -605,14 +605,34 @@ whichEith originalFp = whichFull
     whichFull fp = do
       (trace . mappend "which " . toTextIgnore) fp >> whichUntraced
       where
-        whichUntraced | isAbsolute fp          = checkFile
-                      | dotSlash splitOnDirs   = checkFile
-                      | length splitOnDirs > 0 = lookupPath  >>= leftPathError
-                      | otherwise              = lookupCache >>= leftPathError
+        whichUntraced | isAbsolute fp             = checkFile
+                      | startsWithDot splitOnDirs = checkFile
+                      | length splitOnDirs > 0    = lookupPath  >>= leftPathError
```

Comment:
I don't understand when the guard `length splitOnDirs > 0` can fail (be `False`)... It seems this happens only when `fp` is `[]`.  But this should not happen, should it?  So is the last case dead?
Here are some experiments:
```
ghci> :t splitDirectories
splitDirectories :: FilePath -> [FilePath]
ghci> splitDirectories "."
["."]
ghci> splitDirectories ""
[]
ghci> splitDirectories "/"
["/"]
ghci> splitDirectories "//"
["//"]
ghci> splitDirectories "///"
["///"]
ghci> splitDirectories "/a"
["/","a"]
ghci> splitDirectories "/a/"
["/","a"]
ghci> splitDirectories "/a/b"
["/","a","b"]
ghci> splitDirectories "c:/a"
["c:","a"]
ghci> splitDirectories "c:a"
["c:a"]
ghci> splitDirectories "c:foo.bat"
["c:foo.bat"]
ghci> splitDirectories "c:\\foo.bat"
["c:\\foo.bat"]
ghci> splitDirectories "c:\\\\foo.bat"
["c:\\\\foo.bat"]
ghci> splitDirectories "c:/foo.bat"
["c:","foo.bat"]
```

Reply:
Which output of `splitDirectories` am I looking at? the one from `system-filepath` or `filepath`? :sweat_smile:

Reply:
Oh sorry, this is the one of `filepath` which is used in latest `shelly`.  But the `system-filepath` one is not much different.
@gregwebs : Can you explain the mystery of the `length splitOnDirs > 0` guard?  You introduced it in https://github.com/gregwebs/Shelly.hs/commit/d4415838b673fae377481dd73a8de1e60f3b56db#r70391227

Reply:
Greg agrees that the guard `length splitOnDirs > 0` is unnecessary (and the clause below can be dropped).

Reply:
Excellent! I will amend this PR with this and the other changes between today and tomorrow. Thanks!